File: samples/ios-demo/LoklokDemo/DevToken.swift
import Foundation
import CryptoKit

/// Dev-only token minting so the demo is self-contained. In a real app the token comes
/// from YOUR backend after login — never ship the signing secret in a client.
enum DevToken {
    static func mint(userId: String, name: String, secret: String = "dev-secret-change-me") -> String {
        func b64(_ data: Data) -> String {
            data.base64EncodedString()
                .replacingOccurrences(of: "+", with: "-")
                .replacingOccurrences(of: "/", with: "_")
                .replacingOccurrences(of: "=", with: "")
        }
        let header = b64(Data(#"{"alg":"HS256","typ":"JWT"}"#.utf8))
        let exp = Int(Date().timeIntervalSince1970) + 60 * 60 * 24
        let payload = b64(Data(#"{"sub":"\#(userId)","name":"\#(name)","exp":\#(exp)}"#.utf8))
        let signingInput = "\(header).\(payload)"
        let key = SymmetricKey(data: Data(secret.utf8))
        let sig = HMAC<SHA256>.authenticationCode(for: Data(signingInput.utf8), using: key)
        return "\(signingInput).\(b64(Data(sig)))"
    }
}

File: samples/android-demo/src/main/kotlin/com/loklok/demo/DevToken.kt
package com.loklok.demo

import android.util.Base64
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * Dev-only token minting so the demo is self-contained. In a real app the token is
 * issued by YOUR backend after the user logs in — never ship the signing secret in a client.
 */
object DevToken {
    fun mint(userId: String, name: String, secret: String = "dev-secret-change-me"): String {
        fun b64(bytes: ByteArray) =
            Base64.encodeToString(bytes, Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP)

        val header = b64("""{"alg":"HS256","typ":"JWT"}""".toByteArray())
        val exp = System.currentTimeMillis() / 1000 + 60 * 60 * 24
        val payload = b64("""{"sub":"$userId","name":"$name","exp":$exp}""".toByteArray())
        val signingInput = "$header.$payload"
        val mac = Mac.getInstance("HmacSHA256").apply {
            init(SecretKeySpec(secret.toByteArray(), "HmacSHA256"))
        }
        val sig = b64(mac.doFinal(signingInput.toByteArray()))
        return "$signingInput.$sig"
    }
}
